core/src/main/scala/org/apache/spark/sql/execution/row/RowDeleteExec.scala

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2017 SnappyData, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License. You
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package org.apache.spark.sql.execution.row
+
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode, ExpressionCanonicalizer}
+import org.apache.spark.sql.catalyst.expressions.{Attribute, BindReferences}
+import org.apache.spark.sql.execution.{SparkPlan, TableMutationExec}
+import org.apache.spark.sql.sources.ConnectionProperties
+import org.apache.spark.sql.store.CodeGeneration
+import org.apache.spark.sql.types.{StructField, StructType}
+
+/**
+ * Generated code plan for deletes in a row table.
+ */
+case class RowDeleteExec(child: SparkPlan, resolvedName: String,
+    tableSchema: StructType, keyColumns: Seq[Attribute],
+    connProps: ConnectionProperties, onExecutor: Boolean)
+    extends TableMutationExec(forUpdate = false) {
+
+  @transient private var result: String = _
+  @transient private var stmt: String = _
+  @transient private var batchCount: String = _
+
+  override protected def doProduce(ctx: CodegenContext): String = {
+    val (initCode, commitCode, endCode) = connectionCodes(ctx)
+    result = ctx.freshName("result")
+    ctx.addMutableState("long", result, s"$result = -1L;")
+    stmt = ctx.freshName("statement")
+    batchCount = ctx.freshName("batchCount")
+    val childProduce = doChildProduce(ctx)
+    s"""
+       |if ($result >= 0L) return;
+       |$initCode
+       |try {
+       |  final java.sql.PreparedStatement $stmt = $connTerm.prepareStatement(
+       |    "${s"DELETE FROM $resolvedName WHERE $whereClauseString"}");
```

Comment:
> We already have SnappySession.update() & delete(). IMO those APIs should be enough for such use cases. ( we may change their underlying execution mechanism)

No those are just hacks that were added in demo release for row tables. They are just a partial break-down of SQL strings and an incomplete one at that since it takes only a WHERE clause and not a general query (e.g. sub-query or whatever all of which is cleanly done with this approach). It is completely out of place with the rest of Dataset API. This is the correct way to do it and those old APIs will be deprecated/removed.

> What we were lacking is if each row's criteria is different from each other e.g. CDC case or to generalize, delete using a dataset as criteria.

Yes, mentioned that for deleteFrom. Can you clarify how you are thinking similar can apply to update? I had earlier proposed an Dataset[UpdatableRow] kind of API for applying updates but now I think that the only case where this is useful is a PUT kind of scenario where only some columns are being put so should be subsumed into the putInto API.

> Regarding inefficiencies won't that be documentation issue to advise users to use
dataSet.select("primary_key_col")

Sure for row tables but doesn't apply for column tables. Maybe when we have indexes. Even so that is different from the use of keyColumns above which is an internal SparkPlan-level requirement that says: "update/delete is only possible if the plan knows how to quickly get to a row that was selected from itself"